src/pybind/mgr/smb/fs.py

```diff
@@ -106,3 +121,86 @@ def resolve_exists(
                 raise NotADirectoryError(volpath)
         log.debug('Verified that %r exists in %r', volpath, volume)
         return volpath
+
+
+class _TTLCache:
+    def __init__(self, maxsize: int = 512, ttl: float = 300.0) -> None:
+        self.cache: Dict[Tuple[str, str, str], Tuple[str, float]] = {}
+        self.maxsize: int = maxsize
+        self.ttl: float = ttl
+
+    def _evict(self) -> None:
+        """Evicts items that have expired or if cache size exceeds maxsize."""
+        current_time: float = time.monotonic()
+        keys_to_evict: list[Tuple[str, str, str]] = [
+            key
+            for key, (_, timestamp) in self.cache.items()
+            if current_time - timestamp > self.ttl
+        ]
+        for key in keys_to_evict:
+            del self.cache[key]
+
+        # Further evict if cache size exceeds maxsize
+        if len(self.cache) > self.maxsize:
+            for key in list(self.cache.keys())[
+                : len(self.cache) - self.maxsize
+            ]:
+                del self.cache[key]
+
+    def get(self, key: Tuple[str, str, str]) -> Optional[str]:
+        """Retrieve item from cache if it exists and is not expired."""
+        self._evict()  # Ensure expired items are removed
+        if key in self.cache:
+            value, _ = self.cache[key]
+            return value
+        return None
+
+    def set(self, key: Tuple[str, str, str], value: str) -> None:
+        """Set item in cache, evicting expired or excess items."""
+        self._evict()  # Ensure expired items are removed
+        self.cache[key] = (value, time.monotonic())
+
+    def clear(self) -> None:
+        """Clear all items in the cache."""
+        self.cache.clear()
+
+    def __len__(self) -> int:
+        """Return the number of items currently in the cache."""
+        return len(self.cache)
+
+
+class CachingCephFSPathResolver(CephFSPathResolver):
+    """
+    A subclass of CephFSPathResolver that adds caching to the resolve method
+    to improve performance by reducing redundant path resolutions.
+
+    This implementation uses a TTL (Time-To-Live) cache rather than an LRU (Least
+    Recently Used) cache. The TTL cache is preferred in this scenario because
+    the validity of cached paths is time-sensitive, and we want to ensure that
+    paths are refreshed after a certain period regardless of access frequency.
+    Rlock can be used to synchronize access to the cache, but that is something
+    not required for now & can be later tested.
+    """
+
+    def __init__(
+        self, mgr: Module_T, client: Optional[CephfsClient] = None
```

Comment:
Let's retain the `*,` separator like the base class, forcing `client` to be a keyword only argument.